src/compas/files/base_reader.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import absolute_import
+from __future__ import division
+
+try:
+    from pathlib import Path
+except ImportError:
+    from pathlib2 import Path
+try:
+    from urllib.request import urlopen
+except ImportError:
+    from urllib2 import urlopen
+
+
+class BaseReader(object):
+    """Base class containing file reading functions for file extension specific readers
+
+    Attributes
+    ----------
+    location : str or pathlib object
+        Path or URL to the file
+    """
+    def __init__(self, location):
+        self.is_url = False
+        self.location = self.check_location(location)
+
+    def read(self, mode='text'):
+        """Returns iterable reading file line by line when given a location,
+           either as pathlib object, string containing path or string containing url
+        Arguments
+        ---------
+        mode : string
+            Treat file as ascii or binary
+        Yields
+        -------
+        string
+            yields next line
+        """
+
+        if mode == 'text':
+
+            # TODO: Handle url's to binary files
+            if self.is_url:
+                resp = urlopen(self.location)
+                for line in resp.read().decode('utf-8').split('\n'):
+                    yield line
+            else:
+                try:
+                    with self.location.open(mode='r') as fh:
+                        for line in fh:
+                            yield line
+                except UnicodeDecodeError:
+                    with self.location.open(mode='r',
+                                            errors='replace',
+                                            newline='\r') as fh:
+                        for line in fh:
+                            yield line
+
+        elif mode == 'binary':
+            raise NotImplementedError
+
+        else:
+            raise ValueError('mode {} not recognized'.format(mode))
+
+    def check_location(self, location):
+        """
+        """
+        if str(location).startswith('http'):
+            self.is_url = True
+            return location
+        elif not isinstance(location, Path):
+            pathobj = Path(location)
+        else:
+            pathobj = location
+
+        if pathobj.exists():
+            return pathobj
+        else:
+            raise IOError('File not found.')
+
+    def check_file_signature(self):
```

Comment:
@gonzalocasas:
> This method is great stuff, but I would make a few changes for clarity, because it took me a while to understand what you're doing here:
> 
> 1. `self.file_signature` is a constant per-class, so I would make that uppercase, eg `self.FILE_SIGNATURE`.
> 2. Docs! This is a pretty significant addition to the public API, so, it really needs to be well documented.
> 3. The first check cannot be passed, there are 3 options: 1) the support for urls is implemented 2) not implemented and raises exception, 3) not implemented, but add a `return` immediately.
> 4. Try to avoid nesting checks. The `else` that contains everything is not needed, because the previous code paths end in error (at least, they should, see previous point)
> 
> Then, two questions:
> 
> 1. Does this work with ascii files (since you open the file in binary mode)?
> 2. Would it be possible to use `urlretrieve` and actually remove al the remote stuff to be a one-time download into temp and handle everything as files?

@tetov:
> Questions:
> 
> 1. Yes!
> 2. This goes against what I thought you wanted, to not load a big file into memory?

Reply:
where is this file signature defined? it doesn't seem to have a default on the base reader and obj reader (for example) doesn't seem to define it either?

Reply:
File signatures:
* ply: `b'ply'`
* stl: ascii has`b'SOLID'`, binary has a header..
* off: `b'OFF'`

Not all formats have them but for the ones that do it could be a good check before trying to load a file.

Needs more testing as well.

Reply:
Reworked method a bit now.
https://github.com/compas-dev/compas/blob/d27f338942d4d5109ad888c80534bd1ca8398193/src/compas/files/base_reader.py#L101

And I think I understand what you meant now @gonzalocasas. Downloading file to tmp and then iterating over it would make the whole function more streamlined.

Do you agree with my reasoning in method docstring? Useful to keep the check for file formats with file signatures?